test/raft/distributed_test.cc

```diff
@@ -530,5 +533,85 @@ struct persistence : public raft::persistence {
     }
 };
 
+struct failure_detector : public raft::failure_detector, public server_set {
+    raft::logical_clock _clock;
+
+    // The set of known servers, used to broadcast heartbeats.
+    // The second element of the pair denotes an expiration time after which we remove the server from this set.
+    // Most servers won't have an expiration time; it is assigned only to servers from which we receive a message
+    // without having them added locally with `add_server`.
```

Comment:
I do not think we ever call is_alive() on such nodes.

Reply:
But do these nodes call `is_alive` on us? (`_known` is for **sending** heartbeats, not for receiving them).

Reply:
`is_alive()` is called exactly twice currently:
* by a leader on each known follower (obviously they should be in the config otherwise a leader will not contact them)
* by a follower to check if a leader is alive and before doing that we check that the leader in the current config

So we never call `is_alive()` on a node that is outside of current config.

Reply:
Cannot we have weird situations like a leader outside the configuration (which is a synonym for "the set of voters" AFAIK) or whatever during configuration changes?

Reply:
Yes, we totally can! My point is that it looks like we never call `is_alive()` on it, so failure detector may not know about it. RPC module have to to be able to deliver messages to it.

Reply:
So are you saying that if we have a server `Y` outside configuration (e.g. a leader during config change) and `X` inside, `rpc::add_server(Y)` will be called on `X`? And vice-versa?


I added this expiration thing because I read this comment in raft_rpc.cc:
```cpp
void raft_rpc::add_server(raft::server_id id, raft::server_info info) {
    // Parse gms::inet_address from server_info
    auto in = ser::as_input_stream(bytes_view(info));
    // Entries explicitly managed via `rpc::add_server` and `rpc::remove_server` should never expire
    // while entries learnt upon receiving an rpc message should be expirable.
    _raft_services.update_address_mapping(id, ser::deserialize(in, boost::type<gms::inet_address>()), false);
}
```
and in raft_services.cc:
```cpp
            // The address learnt from a probably unknown server should
            // eventually expire
            self.update_address_mapping(from, std::move(addr), true);
```

so at least in address mappings we have expirable servers for which `add_server` was supposedly not called.

What would be the example of that happening?
cc @ManManson could you explain?

Reply:
> So are you saying that if we have a server `Y` outside configuration (e.g. a leader during config change) and `X` inside, `rpc::add_server(Y)` will be called on `X`? And vice-versa?
No. That is not what I am saying. What I do say is that in this case is_alive(Y) will not be called on X (at least with current code).

> 
> I added this expiration thing because I read this comment in raft_rpc.cc:
> 
> ```c++
> void raft_rpc::add_server(raft::server_id id, raft::server_info info) {
>     // Parse gms::inet_address from server_info
>     auto in = ser::as_input_stream(bytes_view(info));
>     // Entries explicitly managed via `rpc::add_server` and `rpc::remove_server` should never expire
>     // while entries learnt upon receiving an rpc message should be expirable.
>     _raft_services.update_address_mapping(id, ser::deserialize(in, boost::type<gms::inet_address>()), false);
> }
> ```
> 
> and in raft_services.cc:
> 
> ```c++
>             // The address learnt from a probably unknown server should
>             // eventually expire
>             self.update_address_mapping(from, std::move(addr), true);
> ```
> 
> so at least in address mappings we have expirable servers for which `add_server` was supposedly not called.
> 
Yes, as I said RPC does have to be able to deliver messages to nodes outside the config. In your example X should be able to send append_reply to Y.